### PR TITLE
Stream Alternates

### PR DESCRIPTION
**Pull request by derekcollison**
This introduces stream alternates as an optional part of StreamInfo.

Alternates provide a priority list of mirrors and the source in relation to where the request originated.
So if a stream S has 2 mirrors, and we receive a user request for the stream info for S, it will include a sorted list of itself, since it could be the closest, and all mirrors, allowing clients to select read replicas if so desired.

Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

File: server/jetstream_cluster.go
Comment (kozlovic) [inline server/jetstream_cluster.go:6074]: Apparently, you need to check for `js != nil` here. In the code above where you invoke this, the `js.clusterInfo()` is immune to nil pointer because it checks. So either call js.streamAlternates when js is not nil, or check for nil here.

File: server/jetstream_cluster.go
Comment (kozlovic) [inline server/jetstream_cluster.go:6115]: Normally, sort.Slice() asks to implement the "less" function. So is ">" here intended?

**Comment by ripienaar** [comment]
Looks pretty good, you say optional part of stream info - but its not something i can request or not its always there, you mean optional as in it's not there post create for example right? Just to be sure we didnt miss some option somewhere

It's nice at initial start but do you anticipate for example a KV bucket would request info regularly to reconfigure itself for new mirros should they become available?

File: server/raft.go
Comment (matthiashanel) [inline server/raft.go:2504]: always true?

**Review by matthiashanel — CHANGES_REQUESTED** [review]

File: server/stream.go
Comment (matthiashanel) [inline server/stream.go:111]: Domain?

**Review by matthiashanel — COMMENTED** [review]
what is the use case for this?

File: server/jetstream_cluster.go
Comment (derekcollison) [reply server/jetstream_cluster.go:6115]: The test makes sure the order is correct.

File: server/stream.go
Comment (derekcollison) [reply server/stream.go:111]: This is to be able to select a read replica that is closer to you than the source.

Comment (derekcollison) [reply server/stream.go:111]: Right now I don't think we can cross accounts or domains.

Comment (matthiashanel) [reply server/stream.go:111]: yes It's possible to source/mirror cross domains

Comment (derekcollison) [reply server/stream.go:111]: ok so you think we should include domain even if the majority of the time it will be the default or the same?

Remember we can always add things, once we do though hard if not impossible to remove them. Hence I like to start small and provide less to start.

**Comment by derekcollison** [comment]
Won't merge until other PRs land and @kozlovic can do his magic to not have all the redundant commits.

**Review by kozlovic — APPROVED** [review]
LGTM

**Comment by kozlovic** [comment]
@derekcollison Rebase done.. Let's make sure we have a green run, but the rebase itself was painless.

File: server/stream.go
Comment (derekcollison) [reply server/stream.go:112]: I added.

**Pull request merged**
